src/shogun/clustering/KMeansLloydImpl.cpp

```diff
@@ -37,107 +49,117 @@ void CKMeansLloydImpl::Lloyd_KMeans(int32_t k, CDistance* distance, int32_t max_
+					/* if weights_set(j)~=0 */
+					if (weights_set[cl_list_i]!=0.0)
+					{
+						vec=lhs->get_feature_vector(i, vlen, vfree);
+
+						for (j=0; j<dimensions; j++)
+						{
+							mus(j, cl_list_i)-=
+								(vec[j]-mus(j, cl_list_i)) / weights_set[cl_list_i];
+						}
+						lhs->free_feature_vector(vec, i, vfree);
+					}
+					else
+					{
+						/*  mus(:,j)=zeros(dimensions,1) ; */
+						for (j=0; j<dimensions; j++)
+							mus(j, cl_list_i)=0;
+					}
+					
+				}
+
+				cl_list[i] = min_cl;
+			}
+		}
+
+		/* Update Step : Calculate new means */
 		if (!fixed_centers)
 		{
+			
 			/* mus=zeros(dimensions, k) ; */
 			mus.zero();
-			for (int32_t i=0; i<XSize; i++)
+			for (int32_t i=0; i<lhs_size; i++)
 			{
-				int32_t Cl=ClList[i];
+				int32_t Cl=cl_list[i];
 
 				vec=lhs->get_feature_vector(i, vlen, vfree);
 
 				for (int32_t j=0; j<dimensions; j++)
-					mus.matrix[Cl*dimensions+j] += vec[j];
-
+					mus(j, Cl) += vec[j];
 				lhs->free_feature_vector(vec, i, vfree);
 			}
-
+		
 			for (int32_t i=0; i<k; i++)
 			{
 				if (weights_set[i]!=0.0)
 				{
 					for (int32_t j=0; j<dimensions; j++)
-						mus.matrix[i*dimensions+j] /= weights_set[i];
-				}
-			}
-		}
-		rhs_mus->copy_feature_matrix(mus);
-		for (int32_t i=0; i<XSize; i++)
-		{
-			/* ks=ceil(rand(1,XSize)*XSize) ; */
-			const int32_t Pat=CMath::random(0, XSize-1);
-			const int32_t ClList_Pat=ClList[Pat];
-			int32_t imini, j;
-			float64_t mini;
-
-			/* compute the distance of this point to all centers */
-			for(int32_t idx_k=0;idx_k<k;idx_k++)
-				dists[idx_k]=distance->distance(Pat,idx_k);
-
-			/* [mini,imini]=min(dists(:,i)) ; */
-			imini=0 ; mini=dists[0];
-			for (j=1; j<k; j++)
-				if (dists[j]<mini)
-				{
-					mini=dists[j];
-					imini=j;
-				}
-
-			if (imini!=ClList_Pat)
-			{
-				changed++;
-
-				/* weights_set(imini) = weights_set(imini) + 1.0 ; */
-				weights_set[imini]+= 1.0;
-				/* weights_set(j)     = weights_set(j)     - 1.0 ; */
-				weights_set[ClList_Pat]-= 1.0;
-
-				vec=lhs->get_feature_vector(Pat, vlen, vfree);
-
-				for (j=0; j<dimensions; j++)
-				{
-					mus.matrix[imini*dimensions+j]-=
-						(vec[j]-mus.matrix[imini*dimensions+j]) / weights_set[imini];
+						mus(j, i) /= weights_set[i];
 				}
-
-				lhs->free_feature_vector(vec, Pat, vfree);
-
-				/* mu_new = mu_old - (x - mu_old)/(n-1) */
-				/* if weights_set(j)~=0 */
-				if (weights_set[ClList_Pat]!=0.0)
-				{
-					vec=lhs->get_feature_vector(Pat, vlen, vfree);
-
-					for (j=0; j<dimensions; j++)
-					{
-						mus.matrix[ClList_Pat*dimensions+j]-=
-								(vec[j]-mus.matrix[ClList_Pat*dimensions+j]) / weights_set[ClList_Pat];
-					}
-					lhs->free_feature_vector(vec, Pat, vfree);
-				}
-				else
-				{
-					/*  mus(:,j)=zeros(dimensions,1) ; */
-					for (j=0; j<dimensions; j++)
-						mus.matrix[ClList_Pat*dimensions+j]=0;
-				}
-
-				/* ClList(i)= imini ; */
-				ClList[Pat] = imini;
 			}
+					
 		}
+		
+		if (iter%1000 == 0)
```

Comment:
can you print 10 times during the max iterations?
I.e. if max iterations is 100, you should print every 10
If it is 1000, you should print every 100